ddsrouter_core/src/cpp/reader/implementations/rtps/PartitionsReader.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file PartitionsReader.cpp
+ */
+
+#include <fastrtps/rtps/RTPSDomain.h>
+#include <fastrtps/rtps/participant/RTPSParticipant.h>
+
+#include <ddsrouter_utils/exception/InconsistencyException.hpp>
+#include <ddsrouter_utils/Log.hpp>
+
+#include <reader/implementations/rtps/PartitionsReader.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace core {
+namespace rtps {
+
+using namespace eprosima::ddsrouter::core::types;
+
+PartitionsReader::PartitionsReader(
+        const ParticipantId& participant_id,
+        const DdsTopic& topic,
+        std::shared_ptr<PayloadPool> payload_pool,
+        fastrtps::rtps::RTPSParticipant* rtps_participant,
+        std::shared_ptr<DiscoveryDatabase> discovery_database)
+    : CommonReader(
+        participant_id, topic, payload_pool, rtps_participant,
+        history_attributes_(topic),
+        reader_attributes_(topic),
+        topic_attributes_(topic),
+        reader_qos_(topic))
+    , discovery_database_(discovery_database)
+{
+}
+
+types::SpecificWriterQoS PartitionsReader::specific_qos_of_writer_(const types::Guid& guid) const
+{
+    return discovery_database_->get_endpoint(guid).specific_qos();
+}
+
+void PartitionsReader::fill_received_data_(
+        fastrtps::rtps::CacheChange_t* received_change,
+        std::unique_ptr<types::DataReceived>& data_to_fill) const noexcept
+{
+    CommonReader::fill_received_data_(received_change, data_to_fill);
+
+    // Find qos of writer
+    try
+    {
+        data_to_fill->qos.writer_qos = specific_qos_of_writer_(data_to_fill->qos.source_guid);
```

Comment:
Ownership will only work when there are also partitions, as this information is only transferred here for this kind of reader.

Reply:
The fix here is to add `||` in `CommonParticipant::create_reader_`.
I would also change the name of PartitionsReader, but I do not know what to use.

Reply:
Go ahead with the renaming.